Repository/PhpcrOdmRepository.php

```diff
@@ -116,4 +121,62 @@ protected function buildCollection(array $documents)
 
         return $collection;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function add($path, $resource)
+    {
+        Assert::notEq('', trim($path, '/'), 'The root directory cannot be created.');
+        Assert::startsWith($path, '/', 'The target path %s is not absolute.');
+
+        $resolvedPath = $this->resolvePath($path);
+        $parentNode = NodeHelper::createPath($this->getManager()->getPhpcrSession(), $resolvedPath);
+        if (!$parentNode instanceof NodeInterface) {
+            throw new InvalidArgumentException('No parent node created for '.$path);
+        }
+
+        if ($resource instanceof ArrayResourceCollection) {
+            /** @var PhpcrOdmResource[] $resource */
+            foreach ($resource as $item) {
+                Assert::notNull($item->getName(), 'The resource needs a name for the creation');
+                $document = $item->getPayload();
+                $document->setName($item->getName());
+                $document->setParent($parentNode);
+                $this->getManager()->persist($document);
+            }
+        } elseif ($resource instanceof CmfResource) {
+            Assert::notNull($resource->getName(), 'The resource needs a name for the creation');
+
+            $document = $resource->getPayload();
+            $document->setName($resource->getName());
+            $document->setParent($parentNode);
+            $this->getManager()->persist($document);
+        }
+
+        $this->getManager()->flush();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function removeResource($sourcePath, &$deleted)
+    {
+        $document = $this->getManager()->find(null, $sourcePath);
+        $children = $this->getManager()->getChildren($document);
+        $deleted += count($children->toArray());
+
+        $this->getManager()->remove($document);
+        $this->getManager()->flush();
+
+        ++$deleted;
```

Comment:
@WouterJ did the same with $moved in his PR on puli. In general i like return instead of changes by reference.